src/lib.rs

```diff
@@ -60,84 +63,69 @@ const RC: [u64; 24] = [
     0x8000000080008081u64, 0x8000000000008080u64, 0x80000001u64, 0x8000000080008008u64
 ];
 
-macro_rules! REPEAT4 {
-    ($e: expr) => ( $e; $e; $e; $e; )
-}
-
-macro_rules! REPEAT5 {
-    ($e: expr) => ( $e; $e; $e; $e; $e; )
-}
-
-macro_rules! REPEAT6 {
-    ($e: expr) => ( $e; $e; $e; $e; $e; $e; )
-}
-
-macro_rules! REPEAT24 {
-    ($e: expr, $s: expr) => (
-        REPEAT6!({ $e; $s; });
-        REPEAT6!({ $e; $s; });
-        REPEAT6!({ $e; $s; });
-        REPEAT5!({ $e; $s; });
-        $e;
-    )
-}
-
-macro_rules! FOR5 {
-    ($v: expr, $s: expr, $e: expr) => {
-        $v = 0;
-        REPEAT4!({
-            $e;
-            $v += $s;
-        });
-        $e;
-    }
-}
-
+#[allow(unused_assignments)]
 /// keccak-f[1600]
 pub fn keccakf(a: &mut [u64; PLEN]) {
-    let mut b: [u64; 5] = [0; 5];
-    let mut t: u64;
-    let mut x: usize;
-    let mut y: usize;
-
-    for i in 0..24 {
-        // Theta
-        FOR5!(x, 1, {
-            b[x] = 0;
-            FOR5!(y, 5, {
-                b[x] ^= a[x + y];
-            });
-        });
-
-        FOR5!(x, 1, {
-            FOR5!(y, 5, {
-                a[y + x] ^= b[(x + 4) % 5] ^ b[(x + 1) % 5].rotate_left(1);
-            });
-        });
-
-        // Rho and pi
-        t = a[1];
-        x = 0;
-        REPEAT24!({
-            b[0] = a[PI[x]];
-            a[PI[x]] = t.rotate_left(RHO[x]);
-        }, {
-            t = b[0];
-            x += 1;
-        });
-
-        // Chi
-        FOR5!(y, 5, {
-            FOR5!(x, 1, {
-                b[x] = a[y + x];
-            });
-            FOR5!(x, 1, {
-                a[y + x] = b[x] ^ ((!b[(x + 1) % 5]) & (b[(x + 2) % 5]));
-            });
-        });
-
-        // Iota
-        a[0] ^= RC[i];
+    let mut arrays: [[u64; 5]; 24] = [[0; 5]; 24];
+
+    unroll! {
+        for i in 0..24 {
+            // Theta
+            unroll! {
+                for x in 0..5 {
+                    arrays[i][x] = 0;
```

Comment:
This looks useless but it gets way slower without it. I tried using `mem::uninitialized` for the initialisation of `arrays` but that also makes it slower, although not by as much as removing this assignment. Optimisers are weird.